### PR TITLE
[39022] [39017] Add install functionality when adding new shops

### PR DESCRIPTION
**Pull request by rleenders**
# Story Reference

[39022](https://app.clubhouse.io/ns8/story/39022)
[39017](https://app.clubhouse.io/ns8/story/39017)
## Summary

Got a twofer: I tweaked the shop dropdown to change routes instead of just reloading the iframe, and added a store_id query param. This allows both the iframe to resize appropriately, and fires off the dashboard instantiation event in magento (which starts the install process)

## Author Checklist

I have added/updated:

- [ ] Tests
- [ ] Documentation
- [ ] Release notes (title of this PR)

## Version Bump

- [ ] Patch (bug fix - backwards compatible)
- [ ] Minor (new functionality - backwards compatible)
- [ ] Major (significant change - not backwards compatible)

### CODE REVIEW

File: Observer/Admin/DashboardInstantiation.php [hunk 1]
```diff
@@ -139,7 +140,7 @@ public function execute(Observer $observer) : void
                 'moduleVersion' => $moduleVersion,
                 'platformVersion' => (string) $this->productMetadata->getVersion()
             ];
-
+            // die(json_encode($installRequestData));
```
Comment (davehensleyns8) [inline Observer/Admin/DashboardInstantiation.php:143]: Leftover from debugging?

File: view/adminhtml/templates/store_select.phtml [hunk 2]
```diff
@@ -43,11 +43,14 @@ $block = $block;
...
         },
       });
+
       var protectClient = Protect.createClient(clientConfig);
       $('#' + containerElementId).html('');
       protectClient.render(requestedPage, orderIncrementId);
     }
 
+
+
```
Comment (davehensleyns8) [inline view/adminhtml/templates/store_select.phtml:53]: This is a lot of whitespace.

File: Helper/Config.php [hunk 3]
```diff
@@ -232,7 +232,7 @@ public function getStoreMetadata(?int $storeId): ?ProtectMetadata
         $metadatas = $this->getStoreMetadatas();
         $storeId = $storeId !== null ? $storeId : self::EMPTY_STORE_ID;
         return isset($metadatas[$storeId]) ? $metadatas[$storeId] : (
-            isset($metadatas[self::EMPTY_STORE_ID]) ? $metadatas[self::EMPTY_STORE_ID] : null
+            isset($metadatas[self::EMPTY_STORE_ID]) && count($metadatas) == 1 ? $metadatas[self::EMPTY_STORE_ID] : null
```
Comment (davehensleyns8) [inline Helper/Config.php:235]: Could use `===` here.

File: Observer/Admin/DashboardInstantiation.php (hunk 1 above)
Comment (rleenders) [reply Observer/Admin/DashboardInstantiation.php:143]: did you ever consider that I just wanted magento to die?

Comment (CTDoriott) [reply Observer/Admin/DashboardInstantiation.php]: Screen Shot 2020-07-31 at 1 24 38 PM

File: Block/Adminhtml/Container.php [hunk 4]
```diff
@@ -14,10 +14,12 @@
 use Magento\Backend\Block\Template\Context;
 use Magento\Framework\App\Request\Http;
 use Magento\Framework\Event\ManagerInterface;
+use Magento\Store\Model\StoreManagerInterface;
```
Comment (davehensleyns8) [inline Block/Adminhtml/Container.php:17]: Can we please keep these alphabetized? (Obviously Magento's style guide has no rule for this the same way the PHP SDK does, but I still think we should try to do it regardless.)

Comment (rleenders) [reply Block/Adminhtml/Container.php:17]: ... they are alphabetized F comes before S

File: view/adminhtml/templates/store_select.phtml [hunk 5]
```diff
@@ -89,13 +93,22 @@ $block = $block;
...
       var store = stores.find(function(store) {
         return store.id === storeId;
       });
-      renderProtectIFrame(store.token);
+      if(storeId !== paramStoreId) {
+        const newUrl = '<?=$block->getUrl('*/*/*')?>' + '?store_id=' + storeId;
+        window.location = newUrl;
+      }else{
+        renderProtectIFrame(store.token);
+      }
```
Comment (rob-ns8) [inline view/adminhtml/templates/store_select.phtml:107]: Non-blocking:

- Need a space after `if` and before `(`
- Need spaces around `else`

File: view/adminhtml/templates/store_select.phtml [hunk 6]
```diff
@@ -89,13 +93,22 @@ $block = $block;
...
     });
+
     $("#ns8-store-select").on("change", function() {
       var storeId = this.value.toString();
       var store = stores.find(function(store) {
         return store.id === storeId;
       });
-      renderProtectIFrame(store.token);
+      if(storeId !== paramStoreId) {
+        const newUrl = '<?=$block->getUrl('*/*/*')?>' + '?store_id=' + storeId;
```
Comment (aldahick) [inline view/adminhtml/templates/store_select.phtml:103]: (non-blocking) would prefer that all the PHP interpolation happens in the same place for ease of understanding

File: view/adminhtml/templates/store_select.phtml [hunk 7]
```diff
@@ -89,13 +93,22 @@ $block = $block;
...
+        window.location = newUrl;
+      }else{
+        renderProtectIFrame(store.token);
+      }
     });
+
+    $("#ns8-store-select").val(paramStoreId)
+
     $("#ns8-store-select").trigger("change");
```
Comment (aldahick) [inline view/adminhtml/templates/store_select.phtml:112]: This shouldn't be necessary anymore with `.val()` call (unless `paramStoreId` is nullable, in which case we probably want to check that before `.val()` call anyways)

File: Block/Adminhtml/Container.php [hunk 8]
```diff
@@ -15,9 +15,11 @@
 use Magento\Framework\App\Request\Http;
 use Magento\Framework\Event\ManagerInterface;
 use Magento\Framework\View\Result\PageFactory;
+use Magento\Store\Model\StoreManagerInterface;
 use NS8\Protect\Helper\Config;
 use NS8\Protect\Helper\Order;
 use NS8\Protect\Helper\Url;
+use NS8\Protect\Helper\Store;
```
Comment (davehensleyns8) [inline Block/Adminhtml/Container.php:22]: Can we please keep these alphabetized?

File: Block/Adminhtml/StoreSelect.php [hunk 9]
```diff
@@ -4,9 +4,11 @@
 
 use Magento\Backend\Block\Template;
 use Magento\Backend\Block\Template\Context;
+use Magento\Framework\App\Request\Http;
+use Magento\Store\Model\StoreManagerInterface;
 use NS8\Protect\Helper\Config;
-use NS8\Protect\Helper\Store;
 use NS8\Protect\Helper\Url;
+use NS8\Protect\Helper\Store;
```
Comment (davehensleyns8) [inline Block/Adminhtml/StoreSelect.php:11]: Seriously though...

File: Helper/Store.php [hunk 10]
```diff
@@ -40,6 +40,22 @@ public function __construct(
         parent::__construct($context);
     }
 
+    /**
+     *  formats a single store into a more succinct and usable array
```
Comment (davehensleyns8) [inline Helper/Store.php:44]: I know this is super duper nitpicky but there's an extra whitespace at the beginning of this description and it should start with a capital letter.

File: Helper/Store.php [hunk 11]
```diff
@@ -92,4 +102,12 @@ public function getUserStores(): array
         $storeManagerDataList = $this->storeManager->getStores();
         return $this->parseStores($storeManagerDataList);
     }
+
+    /**
+     * get the current store from the storeManager
```
Comment (davehensleyns8) [inline Helper/Store.php:107]: Method description should start with a capital letter, and docblock should have a \@return

File: Observer/Admin/DashboardInstantiation.php [hunk 12]
```diff
@@ -119,13 +119,15 @@ public function __construct(
     public function execute(Observer $observer) : void
     {
         try {
-            $store = $this->storeManager->getStore();
-
-            // TODO use $storeId
-            if ($this->config->isMerchantActive() || $this->config->getAccessToken()) {
+            $event = $observer->getEvent()->getData();
+            $store = $this->storeManager->getStore($event['storeId']);
+            if ($store === null) {
```
Comment (CTDoriott) [inline Observer/Admin/DashboardInstantiation.php:124]: For code cleanliness given this file has a fair amount of conditionals for the number of lines, would you be open to rewriting the store check as:
```
$store = $this->storeManager->getStore($event['storeId']) ?? $this->storeManager->getStore();
```

**Pull request merged**